src/model_theory/fraisse.lean

```diff
@@ -219,23 +220,89 @@ begin
   { exact (hFP _ n).some }
 end
 
-variable (K)
+theorem exists_countable_is_age_of_iff [L.countable_functions] :
+  (∃ (M : bundled.{w} L.Structure), (univ : set M).countable ∧ L.age M = K) ↔
+    K.nonempty ∧
+    (∀ (M N : bundled.{w} L.Structure), nonempty (M ≃[L] N) → (M ∈ K ↔ N ∈ K)) ∧
+    (quotient.mk '' K).countable ∧
+    (∀ (M : bundled.{w} L.Structure), M ∈ K → Structure.fg L M) ∧
+    hereditary K ∧
+    joint_embedding K :=
+begin
+  split,
+  { rintros ⟨M, h1, h2, rfl⟩,
+    resetI,
+    refine ⟨age.nonempty M, age.is_equiv_invariant L M, age.countable_quotient M h1, λ N hN, hN.1,
+      age.hereditary M, age.joint_embedding M⟩, },
+  { rintros ⟨Kn, eqinv, cq, hfg, hp, jep⟩,
+    obtain ⟨M, hM, rfl⟩ := exists_cg_is_age_of Kn eqinv cq hfg hp jep,
+    haveI := ((Structure.cg_iff_countable).1 hM).some,
+    refine ⟨M, countable_encodable _, rfl⟩, }
+end
+
+variables {K} (L) (M)
+
+/-- A structure `M` is ultrahomogeneous if every embedding of a finitely generated substructure
+into `M` extends to an automorphism of `M`. -/
+def is_ultrahomogeneous : Prop :=
+∀ (S : L.substructure M) (hs : S.fg) (f : S ↪[L] M),
+  ∃ (g : M ≃[L] M), f = g.to_embedding.comp S.subtype
+
+variables {L} (K)
+
+/-- A structure `M` is a Fraïssé limit for a class `K` if it is countably generated,
+ultrahomogeneous, and has age `K`. -/
+class is_fraisse_limit : Prop :=
```

Comment:
This might not be a great class. For example, suppose you want to prove `(univ : set M).countable` using `is_fraisse_limit.countable`, now type-class inference will need to find `is_fraisse_limit (M ?K)`, which is problematic if there are any recursive instances.

Reply:
It's now a `structure`.